iotdb-core/consensus/src/main/java/org/apache/iotdb/consensus/pipe/PipeConsensus.java

```diff
@@ -146,33 +147,61 @@ private void initAndRecover() throws IOException {
-                            path.toString(),
-                            new ArrayList<>(),
-                            config,
-                            consensusPipeManager,
-                            syncClientManager);
-                    stateMachineMap.put(consensusGroupId, consensus);
-                    consensus.start(true);
-                  }
-                } catch (Exception e) {
-                  LOGGER.error("Failed to recover consensus from {}", storageDir, e);
-                }
-              })
-          .exceptionally(
-              e -> {
-                LOGGER.error("Failed to recover consensus from {}", storageDir, e);
-                return null;
-              });
+      CompletableFuture<Void> future =
+          CompletableFuture.runAsync(
+                  () -> {
+                    try (DirectoryStream<Path> stream =
+                        Files.newDirectoryStream(storageDir.toPath())) {
+                      for (Path path : stream) {
+                        ConsensusGroupId consensusGroupId =
+                            parsePeerFileName(path.getFileName().toString());
+                        PipeConsensusServerImpl consensus =
+                            new PipeConsensusServerImpl(
+                                new Peer(consensusGroupId, thisNodeId, thisNode),
+                                registry.apply(consensusGroupId),
+                                path.toString(),
+                                new ArrayList<>(),
+                                config,
+                                consensusPipeManager,
+                                syncClientManager);
+                        stateMachineMap.put(consensusGroupId, consensus);
+                        consensus.start(true);
+                      }
+                    } catch (Exception e) {
+                      LOGGER.error("Failed to recover consensus from {}", storageDir, e);
+                    }
+                  })
+              .exceptionally(
+                  e -> {
+                    LOGGER.error("Failed to recover consensus from {}", storageDir, e);
+                    return null;
+                  });
+
+      if (correctPeerListBeforeStart != null) {
+        try {
+          future.get();
+        } catch (Exception e) {
+          throw new RuntimeException(e);
+        }
+        BiConsumer<ConsensusGroupId, List<Peer>> resetPeerListWithoutThrow =
+            (consensusGroupId, peers) -> {
+              try {
+                resetPeerList(consensusGroupId, peers);
+              } catch (ConsensusGroupNotExistException ignore) {
+
+              } catch (Exception e) {
+                LOGGER.warn("Failed to reset peer list while start", e);
+              }
+            };
+        // make peers which are in list correct
+        correctPeerListBeforeStart.forEach(resetPeerListWithoutThrow);
+        // clear peers which are not in the list
+        stateMachineMap.keySet().stream()
```

Comment:
same

Reply:
Indeed. Copy to a new list to avoid concurrent modification.